src/main/java/fi/aalto/cs/apluscourses/presentation/ProgressViewModel.java

```diff
@@ -28,46 +31,49 @@ public ProgressViewModel() {
    */
   public Progress start(int maxValue, String label, boolean indeterminate) {
     var progress = new Progress(maxValue, label, indeterminate);
-    progresses.add(progress);
+    synchronized (lock) {
+      progresses.add(progress);
+    }
     this.updateValues();
+    progress.updated.addListener(this, ProgressViewModel::updateValues);
     return progress;
   }
 
-  public void stop(Progress progress) {
-    progresses.remove(progress);
-    this.updateValues();
-  }
-
+  /**
+   * Stops all Progresses.
+   */
   public void stopAll() {
-    progresses.clear();
+    synchronized (lock) {
+      progresses.clear();
+    }
     this.updateValues();
   }
 
   private void updateValues() {
-    if (progresses.isEmpty()) {
-      this.maxValue.set(0);
-      this.value.set(0);
-      this.indeterminate.set(false);
-    } else {
-      var progress = progresses.getLast();
-      this.value.set(progress.getValue());
-      this.maxValue.set(progress.getMaxValue());
-      this.label.set(progress.getLabel());
-      this.indeterminate.set(progress.getIndeterminate());
+    synchronized (lock) {
+      List<Progress> toBeRemoved = new ArrayList<>();
+      for (var progress : progresses) {
+        if (progress.getValue() >= progress.getMaxValue()) {
+          progress.updated.removeCallback(this);
+          toBeRemoved.add(progress);
+        }
+      }
+      toBeRemoved.forEach(progresses::remove);
```

Comment:
If it feels terrifying to add this much logic to a method, you can make a helper method to CollectionUtil:
```java
public static <T> void removeIf(Collection<T> collection,
                                Predicate<T> predicate,
                                Consumer<T> callbackForRemoved) {
  //...
}
```
which you can unit test with e.g. strings.

Then you can use that here like
```java
CollectionUtil.removeIf(progresses, Progress::isFinished, this::unregisterProgress);
```
where `unregisterProgress` is a simple private method that calls `progress.updated.removeCallback(this)` (or use lambda).